packages/template-retail-react-app/app/hooks/use-address-product-management.js

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2025, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {useState, useEffect, useCallback, useMemo, useRef} from 'react'
+import {nanoid} from 'nanoid'
+import {useCurrentCustomer} from '@salesforce/retail-react-app/app/hooks/use-current-customer'
+import {
+    areAddressesEqual,
+    isAddressEmpty
+} from '@salesforce/retail-react-app/app/utils/address-utils'
+import {isPickupMethod} from '@salesforce/retail-react-app/app/utils/shipment-utils'
+
+/**
+ * Managing address selection state with product delivery items
+ */
+export const useAddressProductManagement = (basket) => {
+    const {data: customer} = useCurrentCustomer()
+
+    const deliveryItems =
+        basket?.productItems?.filter((item) => {
+            const shipment = basket?.shipments?.find((s) => s.shipmentId === item.shipmentId)
+            return !isPickupMethod(shipment?.shippingMethod)
+        }) || []
```

Comment:
this data run on every render, which will trigger almost every methods in this hook.